mtest/libmscore/all_elements/tst_tree_model.cpp

```diff
@@ -0,0 +1,109 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#include <QtTest/QtTest>
+#include "mtest/testutils.h"
+#include "libmscore/score.h"
+
+#define DIR QString("libmscore/all_elements/")
+
+using namespace Ms;
+
+//---------------------------------------------------------
+//   TestTreeModel
+///   Ensures that the tree model is consistent. Starting
+///   from Score each element in the tree should have the
+///   correct parent element, the one whose children list
+///   it appears in.
+//---------------------------------------------------------
+
+class TestTreeModel : public QObject, public MTest
+{
+    Q_OBJECT
+
+    MasterScore * score;
+    void beam(const char* path);
+    void tstTree(QString file);
+    void traverseTree(ScoreElement* element);
+
+    QString elementToText(ScoreElement* element);
+
+private slots:
+    void initTestCase();
+    void tstTreeElements()  { tstTree("layout_elements.mscx");     }
+    void tstTreeTablature() { tstTree("layout_elements_tab.mscx"); }
+    void tstTreeMoonlight() { tstTree("moonlight.mscx");           }
+    void tstTreeGoldberg()  { tstTree("goldberg.mscx");            }
+};
+
+//---------------------------------------------------------
+//   initTestCase
+//---------------------------------------------------------
+
+void TestTreeModel::initTestCase()
+{
+    initMTest();
+}
+
+//---------------------------------------------------------
+//   tstTree
+//---------------------------------------------------------
+
+void TestTreeModel::tstTree(QString file)
+{
+    MasterScore* score = readScore(DIR + file);
+    traverseTree(score);
+}
+
+//---------------------------------------------------------
+//   traverseTree
+///   Checks whether parent element of current element is
+///   correct, then recursively checks all children.
+//---------------------------------------------------------
+
+void TestTreeModel::traverseTree(ScoreElement* element)
+{
+    for (ScoreElement* child : (*element)) {
+        if (child) {
```

Comment:
My preference is to exit early if you can and save yourself an indentation level later on.

```c++
if (!child) {
    continue;
}

// back to previous indentation level
```

But this is a matter of personal taste so do what you think it best.